components/apimgt/org.wso2.carbon.apimgt.rest.api.util/src/main/java/org/wso2/carbon/apimgt/rest/api/util/impl/OAuthJwtAuthenticatorImpl.java

```diff
@@ -0,0 +1,349 @@
+        String accessToken = RestApiUtil.extractOAuthAccessTokenFromMessage(message,
+                RestApiConstants.REGEX_BEARER_PATTERN, RestApiConstants.AUTH_HEADER_NAME);
+
+        if (StringUtils.countMatches(accessToken, APIConstants.DOT) != 2) {
+            log.error("Invalid JWT token. The expected token format is <header.payload.signature>");
+            return false;
+        }
+        try {
+            SignedJWTInfo signedJWTInfo = getSignedJwt(accessToken);
+            String jwtTokenIdentifier = getJWTTokenIdentifier(signedJWTInfo);
+            String jwtHeader = signedJWTInfo.getSignedJWT().getHeader().toString();
+            URL basePath = new URL(message.get(APIConstants.BASE_PATH).toString());
+
+            //Validate token
+            log.info("Starting JWT token validation");
+            JWTValidationInfo jwtValidationInfo =
+                    validateJWTToken(signedJWTInfo, jwtTokenIdentifier, accessToken, jwtHeader, basePath);
+            if (jwtValidationInfo != null) {
+                if (jwtValidationInfo.isValid()) {
+                    if (isRESTApiTokenCacheEnabled) {
+                        getRESTAPITokenCache().put(jwtTokenIdentifier, jwtValidationInfo);
+                    }
+                    //Validating scopes
+                    log.info("JWT token validation success. starting scope validation");
+                    return handleScopeValidation(message, signedJWTInfo, accessToken);
+                } else {
+                    log.error("Invalid JWT token :" + getMaskedToken(jwtHeader));
+                    return false;
+                }
+
+            } else {
+                log.error("Invalid JWT token :" + getMaskedToken(jwtHeader));
+                return false;
+            }
+        } catch (ParseException e) {
+            log.error("Not a JWT token. Failed to decode the token. Reason: " + e.getMessage());
+        } catch (MalformedURLException e) {
+            log.error("Malformed URL found in request path.Reason: " + e.getMessage());
+        }
+        return false;
+    }
+
+    /**
+     * Handle scope validation
+     *
+     * @param accessToken   JWT token
+     * @param signedJWTInfo : Signed token info
+     * @param message       : cxf Message
+     */
+    private boolean handleScopeValidation(Message message, SignedJWTInfo signedJWTInfo, String accessToken)
+            throws ParseException {
+
+        OAuthTokenInfo oauthTokenInfo = new OAuthTokenInfo();
+        oauthTokenInfo.setAccessToken(accessToken);
+        oauthTokenInfo.setEndUserName(signedJWTInfo.getJwtClaimsSet().getSubject());
+
+        if (signedJWTInfo.getJwtClaimsSet().getClaim(APIConstants.JwtTokenConstants.SCOPE) != null) {
+
+            String orgId = message.get(RestApiConstants.ORGANIZATION).toString();
+            String[] scopes = signedJWTInfo.getJwtClaimsSet().getStringClaim(JwtTokenConstants.SCOPE)
+                    .split(JwtTokenConstants.SCOPE_DELIMITER);
```

Comment:
Won't it be a problem if signedJWTInfo.getJwtClaimsSet().getStringClaim(JwtTokenConstants.SCOPE) becomes null when scopes are not present? Shall we do a null check?

Reply:
+1 to put null check

Reply:
We do the null check before coming to this step. 
`  if (signedJWTInfo.getJwtClaimsSet().getClaim(APIConstants.JwtTokenConstants.SCOPE) != null) `
If the scopes in the token are null then we won't proceed. So do we need to add another null check here as well ?